### DISCUSSION TITLE
Resizing a grid panel without shifting layout

### DISCUSSION THREAD

**Discussion by parzival418** [mathuo/dockview#345] [Q&A] [answered]
So I implemented a hotkey that will trigger a side panel to turn to zero using an animation loop.  I added a right panel that can be dragged out by a user.  However my left panel shrinks, and the one on the right grows.  I would like the right panel to stay the same and have the middle one only expand.

Is there any way for me to achieve this? I am trying to do it with listener events between panels but it is super janky and not smooth.

Attached is a video of the behaviour I want to prevent from happening:

<video src="https://github.com/mathuo/dockview/assets/2397603/75985450-6240-4671-ad5c-43209563b04f"/>

**Comment by mathuo** [comment]
Are the two side panels additional dockview panels with hidden headers or part of a seperate gridview control?

**Reply by parzival418** [reply]
They are part of a separate grid view.

There is a root gridView for the overall layout.  One of those panels is my main panel within which I render a dockview component for the main tabs. These tabs can render various workspaces, one of which uses another nested dockview for the different windows in the tabbed workspace.

On a key command I am triggering off an animation loop which animates the `setSize` for the file drawer panel to 0.

**Reply by mathuo** [reply]
> The `LayoutPriority` and `snap` functionalities are not well documented  so I will improve the documenation around these two features, there is also a small bug here mainly before it hasn't been widely used which I will get fixed before it will work as expected but below is the explanation...

Here (https://codesandbox.io/s/ide-example-22wkl6) is an example running from the attached PR which has a minor bug fix but I believe covers the case you are describing? Any questions do let me know.

When you progmatically resize an item the other items needs to resize to fit the remaining space which is why you get the layout shifting 
(the view with 0 width sees that it could grow since new space has become available). This isn't always the desired effect, as in your case.

dockview does have a way of dealing with exactly which views you want grow (in your case you want the middle view to grow and the side view to remain zero) which is a `LayoutPriority` flag.

For example if you was to set the middle content with `LayoutPriority.High` this would indicate that it will resize taking any available space first before any other views which in your case would mean the middle content resizing and preventing the sidebar from resizing to a larger number than zero.

There are two approaches to resizing a view to zero:

```
// set the view to be zero
api.setView({width: 0});

// or set the view to have visibility false
api.setVisible(false);
```

They both broadly function in the same way but if you are hiding a view then `setVisible(false)` is preferred.

`snap: true` indicates that if the view becomes less than the minimum then it will automatically hide that view, you don't need this flag but it can be a nice step. 

```
        event.api.fromJSON({
            grid: {
                height: 1000,
                width: 1000,
                orientation: Orientation.HORIZONTAL,
                root: {
                    type: 'branch',
                    data: [
                        {
                            type: 'leaf',
                            data: {
                                id: 'left-sidebar-id',
                                component: 'left-sidebar',
                                snap: true,
                                minimumWidth: 100,
                            },
                            size: 200,
                        },
                        {
                            type: 'leaf',
                            data: {
                                id: 'middle-content-id',
                                component: 'middle-content',
                                priority: LayoutPriority.High,
                                minimumWidth: 100,
                            },
                            size: 600,
                        },
                        {
                            type: 'leaf',
                            data: {
                                id: 'right-sidebar-id',
                                component: 'right-sidebar',
                                snap: true,
                                minimumWidth: 100,
                            },
                            size: 200,
                        },
                    ],
                },
            },
        });
```

**Reply by parzival418** [reply]
This is amazing!  Will let you know when I have implemented it!  Thanks for being so responsive.  This library just keeps getting better.This is amazing!  Will let you know when I have implemented it!  Thanks for being so responsive.  This library just keeps getting better.

**Reply by parzival418** [reply]
Quick question.  Are these properties like snap and priority things I can set from the panel api inside the components programatically? Or only on the layout json?

**Reply by mathuo** [reply]
You can include both `snap` and `priority` when adding items to the Gridview through `.addPanel()`. Those flags don't exist on the Dockview since they don't make sense in that context. I will make sure I add all this to the gridview docs.

**Reply by mathuo** [reply]
Both this change and the floating group issues are currently available in an experimental build found [here](https://github.com/mathuo/dockview/issues/349). An official build will follow soon.

**Reply by mathuo** [reply, marked as answer]
Released in version `1.8.4`
